lib/handler/file.c

```diff
@@ -110,6 +119,46 @@ static void do_proceed(h2o_generator_t *_self, h2o_req_t *req)
         do_close(&self->super, req);
 }
 
+static void do_multirange_proceed(h2o_generator_t *_self, h2o_req_t *req) {
+    struct st_h2o_sendfile_generator_t *self = (void *)_self;
+    size_t rlen, *range_cur = self->range_infos + 2 * self->finished_range, used_buf = 0;
+    ssize_t rret;
+    h2o_iovec_t vec;
+
+    if (self->bytesleft == 28) {
+        vec.len = sprintf(self->buf, "\r\n--%s--\r\n", self->boundary.base);
+        vec.base = self->buf;
+        h2o_send(req, &vec, 1, 1);
+        do_close(&self->super, req);
+        return;
```

Comment:
Just to make sure I did not misunderstand, is this part what you meant "polishing up the push generator"? TBO I don't fully understand how `do_procceed()` work when file is larger than `buf`, though I just followed that function and write `do_multirange_proceed()`. What is the process after non-finished sending, i.e. when leave the proceed function after calling h2o_send with `is_finished` 0? Sorry for not fully understanding the workflow, I tried my best...